File: shared/modules/game.ts
import { z } from 'zod';
import { Scheme as KindCodeScheme } from '../constants/kindCode';
import { FieldOptsScheme } from '../constants/fieldOpts';
import { teamCodeSchema } from "../constants/teams";

const DatetimeScheme = z.string().datetime({ offset: true });

const CPBLGameId = z.string(); // TODO check with correct id regex
const PlayerId = z.string();
const TeamId = z.string();
const Url = z.string().url();
const gameSeason = z.enum([
  '', // 短期賽，沒有分季
  '0', // 單一賽季
  '1', // 上半季
  '2', // 下半季
]);

export type GameSeason = z.infer<typeof gameSeason>;

const ResultScheme = z.enum([
  'pending', // 未結束,
  'ended', // 結束,
  'postponed', // 延賽,
  'suspended', // 保留比賽
]);

export type Result = z.infer<typeof ResultScheme>;

type LinkResource = {
  type: 'link';
  src: z.infer<typeof Url>;
};

export const GameScheme = z.object({
  id: CPBLGameId,
  gameNo: z.number(),
  year: z.string(),
  gameKindCode: KindCodeScheme,
  gameSeason: gameSeason,
  gameSeriesNo: z.number(),
  isGameStop: z.boolean(),
  startDatetime: DatetimeScheme,
  endDatetime: DatetimeScheme.nullable(),
  durationSeconds: z.number(),
  field: FieldOptsScheme,
  result: ResultScheme,
  homeScore: z.number(),
  visitingScore: z.number(),
  reserveDate: DatetimeScheme.nullable(),
  homeTeamCode: teamCodeSchema,
  homeTeamName: z.string(),
  homeTeamIconUrl: Url,
  visitingTeamCode: teamCodeSchema,
  visitingTeamName: z.string(),
  visitingTeamIconUrl: Url,
  mvpPlayerId: PlayerId.or(z.literal('')),
  mvpPlayerName: z.string(),
  mvpCount: z.number().nullable(),
  visitingPitcherId: PlayerId.or(z.literal('')),
  visitingPitcherName: z.string(),
  homePitcherId: PlayerId.or(z.literal('')),
  homePitcherName: z.string(),
  winningPitcherId: PlayerId,
  winningPitcherName: z.string(),
  loserPitcherId: PlayerId.or(z.literal('')),
  loserPitcherName: z.string(),
  closerId: PlayerId.or(z.literal('')),
  closerName: z.string(),
});

export type Game = z.infer<typeof GameScheme>;

File: jumbotron-flutter/lib/shared/models/game/game.dart
import 'package:bandwagon/shared/utils/time.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:bandwagon/shared/constants/game.dart';

part 'game.g.dart';

@JsonSerializable()
class Game {
  final String id;
  final int gameNo;
  final String year;
  final GameKind gameKindCode;
  final String gameSeason;
  final int gameSeriesNo;
  final bool isGameStop;
  final bool isPlayBall;

  final int durationSeconds;
  final String field;
  final GameResult result;
  final int homeScore;
  final int visitingScore;

  final String homeTeamCode;
  final String homeTeamName;
  final String homeTeamIconUrl;
  final String visitingTeamCode;
  final String visitingTeamName;
  final String visitingTeamIconUrl;
  final String mvpPlayerId;
  final String mvpPlayerName;
  final int? mvpCount;
  final String visitingPitcherId;
  final String visitingPitcherName;
  final String homePitcherId;
  final String homePitcherName;
  final String winningPitcherId;
  final String winningPitcherName;
  final String loserPitcherId;
  final String loserPitcherName;
  final String closerId;
  final String closerName;

  @RFC3339TimeConverter()
  final DateTime startDatetime;

  @RFC3339TimeConverter()
  final DateTime? endDatetime;

  @RFC3339TimeConverter()
  final DateTime? reserveDate;

  Game({
    required this.id,
    required this.gameNo,
    required this.year,
    required this.gameKindCode,
    required this.gameSeason,
    required this.gameSeriesNo,
    required this.isGameStop,
    required this.isPlayBall,
    required this.startDatetime,
    this.endDatetime,
    required this.durationSeconds,
    required this.field,
    required this.result,
    required this.homeScore,
    required this.visitingScore,
    this.reserveDate,
    required this.homeTeamCode,
    required this.homeTeamName,
    required this.homeTeamIconUrl,
    required this.visitingTeamCode,
    required this.visitingTeamName,
    required this.visitingTeamIconUrl,
    required this.mvpPlayerId,
    required this.mvpPlayerName,
    this.mvpCount,
    required this.visitingPitcherId,
    required this.visitingPitcherName,
    required this.homePitcherId,
    required this.homePitcherName,
    required this.winningPitcherId,
    required this.winningPitcherName,
    required this.loserPitcherId,
    required this.loserPitcherName,
    required this.closerId,
    required this.closerName,
  });

  factory Game.fromJson(Map<String, dynamic> json) => _$GameFromJson(json);
  Map<String, dynamic> toJson() => _$GameToJson(this);
}
